### PR TITLE
[YoutubeDL] simultaneous download for segments 

### PR DESCRIPTION
**Pull request by nao20010128nao**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

This allows to download 2 or more segmented formats at once. This is not used by any other FDs for now, but is required by #888 

image

patch for testing

```diff
diff --git a/mline.py b/mline.py
new file mode 100644
index 000000000..aa17986c5
--- /dev/null
+++ b/mline.py
@@ -0,0 +1,18 @@
+from yt_dlp.minicurses import MultilinePrinter
+
+
+import sys
+import itertools
+import time
+
+from yt_dlp.minicurses import MultilinePrinter
+
+with MultilinePrinter(sys.stdout, 7) as mp:
+    for i in itertools.count():
+        fb = ''
+        if i % 3 == 0:
+            fb += 'Fizz'
+        if i % 5 == 0:
+            fb += 'Buzz'
+        mp.print_at_line(fb, i % 7)
+        time.sleep(0.1)
diff --git a/yt_dlp/YoutubeDL.py b/yt_dlp/YoutubeDL.py
index 50e902c53..c6c05a897 100644
--- a/yt_dlp/YoutubeDL.py
+++ b/yt_dlp/YoutubeDL.py
@@ -127,6 +127,7 @@ from .extractor import (
 )
 from .extractor.openload import PhantomJSwrapper
 from .downloader import (
+    DashSegmentsFD,
     FFmpegFD,
     get_suitable_downloader,
     shorten_protocol_name
@@ -2760,6 +2761,8 @@ class YoutubeDL(object):
                     if len(_protocols) == 1:  # All requested formats have same protocol
                         info_dict['protocol'] = _protocols.pop()
                     directly_mergable = FFmpegFD.can_merge_formats(info_dict, self.params)
+                    downloaded = []
+                    merger = FFmpegMergerPP(self)
                     if dl_filename is not None:
                         self.report_file_already_downloaded(dl_filename)
                     elif (directly_mergable and get_suitable_downloader(
@@ -2767,9 +2770,19 @@ class YoutubeDL(object):
                         info_dict['url'] = '\n'.join(f['url'] for f in requested_formats)
                         success, real_download = self.dl(temp_filename, info_dict)
                         info_dict['__real_download'] = real_download
+                    elif (len(requested_formats) > 1 and get_suitable_downloader(
+                            info_dict, self.params) == DashSegmentsFD):
+                        for f in requested_formats:
+                            f['filepath'] = fname = prepend_extension(
+                                correct_ext(temp_filename, info_dict['ext']),
+                                'f%s' % f['format_id'], info_dict['ext'])
+                            downloaded.append(f['filepath'])
+                            if not self._ensure_dir_exists(fname):
+                                return
+                        info_dict['url'] = requested_formats[0]['url']
+                        success, real_download = self.dl(temp_filename, info_dict)
+                        info_dict['__real_download'] = real_download
                     else:
-                        downloaded = []
-                        merger = FFmpegMergerPP(self)
                         if self.params.get('allow_unplayable_formats'):
                             self.report_warning(
                                 'You have requested merging of multiple formats '
@@ -2802,14 +2815,14 @@ class YoutubeDL(object):
                             partial_success, real_download = self.dl(fname, new_info)
                             info_dict['__real_download'] = info_dict['__real_download'] or real_download
                             success = success and partial_success
-                        if merger.available and not self.params.get('allow_unplayable_formats'):
-                            info_dict['__postprocessors'].append(merger)
-                            info_dict['__files_to_merge'] = downloaded
-                            # Even if there were no downloads, it is being merged only now
-                            info_dict['__real_download'] = True
-                        else:
-                            for file in downloaded:
-                                files_to_move[file] = None
+                    if merger.available and not self.params.get('allow_unplayable_formats'):
+                        info_dict['__postprocessors'].append(merger)
+                        info_dict['__files_to_merge'] = downloaded
+                        # Even if there were no downloads, it is being merged only now
+                        info_dict['__real_download'] = True
+                    else:
+                        for file in downloaded:
+                            files_to_move[file] = None
                 else:
                     # Just a single file
                     dl_filename = existing_file(full_filename, temp_filename)
diff --git a/yt_dlp/downloader/dash.py b/yt_dlp/downloader/dash.py
index 734eab3ef..032733c10 100644
--- a/yt_dlp/downloader/dash.py
+++ b/yt_dlp/downloader/dash.py
@@ -18,46 +18,40 @@ class DashSegmentsFD(FragmentFD):
         if info_dict.get('is_live'):
             self.report_error('Live DASH videos are not supported')
 
-        fragment_base_url = info_dict.get('fragment_base_url')
-        fragments = info_dict['fragments'][:1] if self.params.get(
-            'test', False) else info_dict['fragments']
-
-        real_downloader = get_suitable_downloader(
-            info_dict, self.params, None, protocol='dash_frag_urls', to_stdout=(filename == '-'))
-
-        ctx = {
-            'filename': filename,
-            'total_frags': len(fragments),
-        }
-
-        if real_downloader:
-            self._prepare_external_frag_download(ctx)
-        else:
-            self._prepare_and_start_frag_download(ctx, info_dict)
-
-        fragments_to_download = []
-        frag_index = 0
-        for i, fragment in enumerate(fragments):
-            frag_index += 1
-            if frag_index <= ctx['fragment_index']:
-                continue
-            fragment_url = fragment.get('url')
-            if not fragment_url:
-                assert fragment_base_url
-                fragment_url = urljoin(fragment_base_url, fragment['path'])
-
-            fragments_to_download.append({
-                'frag_index': frag_index,
-                'index': i,
-                'url': fragment_url,
-            })
-
-        if real_downloader:
-            self.to_screen(
-                '[%s] Fragment downloads will be delegated to %s' % (self.FD_NAME, real_downloader.get_basename()))
-            info_copy = info_dict.copy()
-            info_copy['fragments'] = fragments_to_download
-            fd = real_downloader(self.ydl, self.params)
-            return fd.real_download(filename, info_copy)
-
-        return self.download_and_append_fragments(ctx, fragments_to_download, info_dict)
+        fmts = info_dict.get('requested_formats', [info_dict])
+
+        args = []
+
+        for idic in fmts:
+            fragment_base_url = idic.get('fragment_base_url')
+            fragments = idic['fragments'][:1] if self.params.get(
+                'test', False) else idic['fragments']
+
+            ctx = {
+                'filename': idic['filepath'],
+                'total_frags': len(fragments),
+                'ctx_id': idic['filepath'],
+            }
+
+            self._prepare_and_start_frag_download(ctx, idic)
+
+            fragments_to_download = []
+            frag_index = 0
+            for i, fragment in enumerate(fragments):
+                frag_index += 1
+                if frag_index <= ctx['fragment_index']:
+                    continue
+                fragment_url = fragment.get('url')
+                if not fragment_url:
+                    assert fragment_base_url
+                    fragment_url = urljoin(fragment_base_url, fragment['path'])
+
+                fragments_to_download.append({
+                    'frag_index': frag_index,
+                    'index': i,
+                    'url': fragment_url,
+                })
+            
+            args.append([ctx, fragments_to_download, info_dict])
+
+        return self.download_and_append_fragments_multiple(*args)
```

### CODE REVIEW

**Comment by nao20010128nao** [comment]
why does this bypass MLP?

image

**Comment by pukkandan** [comment]
can u make the existing FDs use this (with a new switch)? or should we deal with that in a separate PR?

File: yt_dlp/downloader/fragment.py [hunk 1]
```diff
@@ -324,7 +336,42 @@ def _prepare_external_frag_download(self, ctx):
             'fragment_index': 0,
         })
 
-    def download_and_append_fragments(self, ctx, fragments, info_dict, *, pack_func=None, finish_func=None):
+    # ctx1, fragments1, info_dict1, ctx2, fragments2, info_dict2, ...
+    # length of arguments must be multiple of 3, or you'll see broken console
+    def download_and_append_fragments_multiple(self, *args, pack_func=None, finish_func=None):
```
Comment (pukkandan) [inline yt_dlp/downloader/fragment.py:341]: not good design. better to let args be (a list of) dicts/tuples

Comment (nao20010128nao) [reply yt_dlp/downloader/fragment.py:341]: okay

**Comment by nao20010128nao** [comment]
@pukkandan 
> can u make the existing FDs use this (with a new switch)? or should we deal with that in a separate PR?
I think that should be done in a separate PR, since this PR is focused on #888

File: yt_dlp/downloader/fragment.py [hunk 2]
```diff
@@ -356,7 +358,8 @@ def __exit__(self, exc_type, exc_val, exc_tb):
                 pass
 
         spins = []
-        for idx, ctx, fragments, info_dict in zip(itertools.count(0), args[::3], args[1::3], args[2::3]):
+        for idx, cfi in enumerate(args):
+            ctx, fragments, info_dict = cfi
```
Comment (pukkandan) [inline yt_dlp/downloader/fragment.py:362]: u can do
```py
        for idx, (ctx, fragments, info_dict) in enumerate(args):
```

**Comment by nao20010128nao** [comment]
`--newline`

image

**Comment by nao20010128nao** [comment]
2, but only stdout is redirected

image

**Comment by nao20010128nao** [comment]
2, but both stdout and stderr
image

**Comment by nao20010128nao** [comment]
note for non-Discord users: the number above corresponds to the following list

```
1. text file
2. tee
3. win 7/8 cmd
4. win 10 cmd
5. powershell
6. unix shell
7. with --newline
8. with --no-progress
```

**Comment by nao20010128nao** [comment]
4.
image

**Comment by nao20010128nao** [comment]
5, but on Windows 10

image

**Comment by nao20010128nao** [comment]
as commit history says, dollar signs on non-TTY mode is now gone away.

**Comment by nao20010128nao** [comment]
3. on Windows 7 (without fallback)

VirtualBox_IE11 - Win7_22_09_2021_19_43_53

**Comment by pukkandan** [comment]
yeah, I expected as much. Nothing we can do about it afaik. Probably best to detect the OS and fallback to --newline style

**Comment by nao20010128nao** [comment]
4, with version check (not fallback)

image

3, with version check (fallback activated)

VirtualBox_IE11 - Win7_22_09_2021_20_07_05

**Comment by nao20010128nao** [comment]
@pukkandan Do I have to test on Windows 8.1 too?

**Comment by nao20010128nao** [comment]
I have pushed all changes. Can merge if there's no problem

File: yt_dlp/downloader/common.py [hunk 3]
```diff
@@ -236,20 +242,36 @@ def report_destination(self, filename):
...
+            if compat_os_name == 'nt' or not sys.stderr.isatty():
                 prev_len = getattr(self, '_report_progress_prev_line_length',
                                    0)
                 if prev_len > len(fullmsg):
                     fullmsg += ' ' * (prev_len - len(fullmsg))
                 self._report_progress_prev_line_length = len(fullmsg)
                 clear_line = '\r'
             else:
-                clear_line = ('\r\x1b[K' if sys.stderr.isatty() else '\r')
+                clear_line = '\r\x1b[K'
```
Comment (pukkandan) [inline yt_dlp/downloader/common.py:274]: are you sure about the change?

File: yt_dlp/downloader/fragment.py [hunk 4]
```diff
@@ -347,7 +359,44 @@ def decrypt_fragment(fragment, frag_content):
 
         return decrypt_fragment
 
-    def download_and_append_fragments(self, ctx, fragments, info_dict, *, pack_func=None, finish_func=None):
+    def download_and_append_fragments_multiple(self, *args, pack_func=None, finish_func=None):
```
Comment (pukkandan) [inline yt_dlp/downloader/fragment.py:362]: I don't like this function name, but I dont have any better ideas either 😅

File: yt_dlp/minicurses.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+try:
+    # curses is in standard modules, but like sqlite in cookies.py,
+    # it is not present under some circumstances (especially when running on Windows)
+    import curses
+    has_curses = True
+    CURSES_ERROR = (curses.error, )
+except ImportError:
+    has_curses = False
+    CURSES_ERROR = (NameError, )
```
Comment (pukkandan) [inline yt_dlp/minicurses.py:14]: isnt it possible to completely get rid of this dependancy by hard coding the UNIX control chars (like for windows)? afaik, all flavors and terminals use the same standard, correct?

**Review by pukkandan — APPROVED** [review]
otherwise looks good. Some refactoring may be needed as we add this to the existing downloaders. But what changes are needed is hard to predict rn. So we'll deal with it later

File: yt_dlp/downloader/common.py (hunk 3 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/common.py:274]: this is to fix when doing `yt-dlp ... | cat` shows broken progress.

File: yt_dlp/downloader/fragment.py (hunk 4 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/fragment.py:362]: the next candidate I came up with is `parallel_download_and_append_fragments`. how about this?

File: yt_dlp/minicurses.py (hunk 5 above)
Comment (nao20010128nao) [reply yt_dlp/minicurses.py:14]: if we declare that "we only support ANSI escape sequence for Unix platforms", we can hardcode it. then it'll check for `TERM` variable on non-Windows platforms.

File: yt_dlp/downloader/fragment.py (hunk 4 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/fragment.py:362]: `download_and_append_fragments` is also parallel (if `-N` is set) then :sweat_smile:

File: yt_dlp/minicurses.py (hunk 5 above)
Comment (nao20010128nao) [reply yt_dlp/minicurses.py:14]: > Terminal emulators for communicating with local programs as well as remote machines and the text system console almost always support ANSI escape codes. 

https://en.wikipedia.org/wiki/ANSI_escape_code

looks like I can do that

Comment (nao20010128nao) [reply yt_dlp/minicurses.py]: done

**Comment by ale5000-git** [comment]
@nao20010128nao

Can you allow also simultaneous download of audio and video please?
Example: `yt-dlp -f bestvideo+bestaudio https://www.youtube.com/watch?v=999tf5-ONGw`

**Comment by pukkandan** [comment]
@ale5000-git This PR is setting up the framework that is needed to implement that in the future

**Comment by nao20010128nao** [comment]
@ale5000-git I have no plan on supporting downloading 2 or more streams in generic cases like this, since Python's multithreading is kind of poor. 
You can do that on your own if you want and then make PR, but you may receive a lot of error reports. Do you wish to do that even with this?

**Comment by ale5000-git** [comment]
Currently I have no time for this, but it is a thing that must be done before or later.
The Youtube example maybe is less important but on some sites where download may take 3 hours and multithreading may change things.

**Pull request merged**